aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/SynchronousNonBlockingStepTest.java

```diff
@@ -0,0 +1,245 @@
+        j.waitForMessage("b branch", b);
+        System.out.println("b branch finishes");
+
+        // Check that "a" branch is effectively blocked
+        j.assertLogNotContains("a branch", b);
+
+        // Notify "a" branch
+        System.out.println("Continue on wait0");
+        SynchronousNonBlockingStep.notify("wait0");
+
+        // Wait for "a" branch to finish
+        j.waitForMessage("a branch", b);
+        j.waitForCompletion(b);
+    }
+
+    public static final class SynchronousNonBlockingStep extends AbstractStepImpl implements Serializable {
+
+        public static final class State {
+            private static State state;
+            static synchronized State get() {
+                if (state == null) {
+                    state = new State();
+                    return state;
+                }
+                return state;
+            }
+            private State() {}
+            final Set<String> started = new HashSet<String>();
+        }
+
+        private String id;
+
+        @DataBoundConstructor
+        public SynchronousNonBlockingStep(String id) {
+            this.id = id;
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        public static void waitForStart(String id, Run<?,?> b) throws IOException, InterruptedException {
+            State s = State.get();
+            synchronized (s) {
+                while (!s.started.contains(id)) {
+                    if (b != null && !b.isBuilding()) {
+                        throw new AssertionError(JenkinsRule.getLog(b));
+                    }
+                    s.wait(1000);
+                }
+            }
+        }
+
+        public static final void notify(String id) {
+            State s = State.get();
+            synchronized (s) {
+                if (s.started.contains(id)) {
+                    s.started.remove(id);
+                    s.notifyAll();
+                }
+            }
+        }
+
+        public static class StepExecutionImpl extends AbstractSynchronousNonBlockingStepExecution<Void> {
+
+            @Inject(optional=true) 
+            private SynchronousNonBlockingStep step;
+
+            @Override
+            protected Void run() throws Exception {
+                System.out.println("Starting syncnonblocking " + step.getId());
+                // Send a test message to the listener
+                getContext().get(TaskListener.class).getLogger().println("Test Sync Message");
+
+                State s = State.get();
+                synchronized (s) {
+                    s.started.add(step.getId());
+                    s.notifyAll();
+                }
+                if (step.getId().equals("waitinterrupt")) {
+                    while (true) {
+                        // Keep in an interruptible state
+                        if (Thread.interrupted()) {
+                            getContext().get(TaskListener.class).getLogger().println("Run thread interrupted");
```

Comment:
Or even have no special handling in the step at all—just let the `InterruptedException` be thrown out of `run` (again what a normal implementation would do). Then the test script can use a `try`-`catch` block to assert that `InterruptedException` was thrown from the step.